### PR TITLE
Update REST-endpoints and their tests

### LINKED ISSUE #40: Update codebase following facade-changes
Following #38 we need to update the following:
- ~~Facade Testing~~ (#36)
- ~~REST~~
- ~~REST-testing~~
- ~~API-doc~~ (95% done)

It should be good for now. During Sprint 3 we may return and implement more features.

### PR DESCRIPTION
**Pull request by Runi-VN**
Fixes #40 (for now)

Following some new methods implemented in #38 (and a few extra) this pull request will update REST-endpoints and their respective tests.

`getPersonByPhone` only returns one result. If there is multiple results an exception will be thrown by EntityManager.
I believe `phone.number` should be unique but it is not by design as far as I can see.

I had a lot of issues with `testGetAllZipCodes`:  
The @BeforeEach does not persist items in the same order everytime, but I think I found a way for the test to never fail, but let me know if it does.

### CODE REVIEW

File: src/main/java/facades/SearchFacade_Impl.java [hunk 1]
```diff
@@ -400,7 +423,7 @@ public PersonDTO_OUT addPersonWithEverything(PersonDTO_IN personDTO) {
 
             // Check if any people live in the city. 
             if (persons == null || persons.isEmpty()) {
-                throw new WebApplicationException("No Persons lives in that city.", 400);
+                throw new WebApplicationException("No Persons lives in that city.", 404);
```
Comment (MalteMagnussen) [inline src/main/java/facades/SearchFacade_Impl.java]: Kan du forklare den her? (Ikke at det er forkert. Jeg forstår bare ikke hvorfor :))

File: src/main/java/rest/SearchResource.java [hunk 2]
```diff
@@ -177,29 +178,24 @@ public PersonDTO_OUT addPersonWithHobbies(PersonDTO_IN person) {
...
                 @ApiResponse(responseCode = "404", description = "No persons in that city found")})
-    public List<PersonDTO_OUT> getPersonsByCity(@QueryParam("city") String city) {
-        //    /api/search/city?city=<city>
-        //example mentions both zip & city?
-        //get from facade
-        List<PersonDTO_OUT> returnList = new ArrayList();
-        return returnList;
+    public List<PersonDTO_OUT> getPersonsByCity(@QueryParam("zip") String zip, @QueryParam("city") String city) {
+        //    /api/search/city?zip=<zip>&city=<city>
+        return FACADE.getPersonsInCity(new CityInfoDTO_IN(zip, city));
```
Comment (MalteMagnussen) [inline src/main/java/rest/SearchResource.java]: Sick that it just worked that smoothly. I was highly doubtful that my methods would work.

File: src/test/java/rest/SearchResourceTest.java [hunk 3]
```diff
@@ -274,6 +294,140 @@ public void testGetPersonByFullName()
...
+        String zip = city1.getZipCode(); //2800
+        String city = city1.getCity(); //Lyngby
+        List<PersonDTO_OUT> result;
+        
+        //Act
+         result = get("search/city?zip={zip}&city={city}", zip, city). then()
+                .assertThat()
+                .statusCode(HttpStatus.OK_200.getStatusCode())
+                .extract().body()
+                .jsonPath().getList(".", PersonDTO_OUT.class); //https://stackoverflow.com/a/53006523
```
Comment (MalteMagnussen) [inline src/test/java/rest/SearchResourceTest.java]: How does this work? Looks cool.

Comment (MalteMagnussen) [reply src/test/java/rest/SearchResourceTest.java]: Ahh, lmao, I see the stackoverflow now. Cool. Good job.

File: src/test/java/rest/SearchResourceTest.java [hunk 4]
```diff
@@ -274,6 +294,140 @@ public void testGetPersonByFullName()
...
+////        System.out.println(result.getClass());
+////        System.out.println(expResult.getClass());
+////        System.out.println(expResult.get(0).getClass());
+////        System.out.println(result.get(0).getClass());
+////        System.out.println(result);
+////        System.out.println(expResult);
+        //assertThat((result), equalTo(expResult));
+        //assertEquals(expResult, result);
+        //assertThat(expResult, Matchers.containsInAnyOrder(result.get(0), result.get(1), result.get(2)));
+        assertThat("ARE THEY EQUAL", result, containsInAnyOrder(expResult.toArray()));
```
Comment (MalteMagnussen) [inline src/test/java/rest/SearchResourceTest.java]: Haha, what? "ARE THEY EQUAL"? What is going on here? How does this line work?

File: src/test/java/rest/SearchResourceTest.java [hunk 5]
```diff
@@ -274,6 +294,140 @@ public void testGetPersonByFullName()
...
+    @Tag("skipBeforeEach")
+    @Test public void testGetAllZipCodes_Exception1(){
+        //Assert
+        given().
+                get("/search/zip/").then().
+                statusCode(HttpStatus.BAD_REQUEST_400.getStatusCode()).
+                assertThat().
+                body("code", equalTo(400)).
+                body("message", equalTo("No cities in the database."));
+    }
```
Comment (MalteMagnussen) [inline src/test/java/rest/SearchResourceTest.java]: Could you explain how this works ? :)

File: src/main/java/facades/SearchFacade_Impl.java (hunk 1 above)
Comment (Runi-VN) [reply src/main/java/facades/SearchFacade_Impl.java]: Det er faktisk ikke nødvendigvis korrekt.

400 Bad Request
*The server cannot or will not process the request due to an apparent client error (e.g., malformed request syntax, size too large, invalid request message framing, or deceptive request routing).*
vs

404 Not Found
*The requested resource could not be found but may be available in the future. Subsequent requests by the client are permissible.*

Da jeg lavede det kiggede jeg kun på navnet **NOT FOUND** -- men nu bliver jeg helt i tvivl om hvilken der er korrekt.

File: src/test/java/rest/SearchResourceTest.java (hunk 4 above)
Comment (Runi-VN) [reply src/test/java/rest/SearchResourceTest.java]: assertThat(String**REASON**, Object**ToCompare**, containsInAnyOrder(Elements\*))

\* I use toArray() to list the all the elements instead of just the list (doesn't work.

Essentially it says

assertThat("message", {2800, 8000, 9000} , containsInAnyOrder([2800, 8000, 9000]\*))

\* Where the iterator (indexPosition) does not matter

Documentation:  http://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/Matchers.html#containsInAnyOrder(java.util.Collection)

File: src/test/java/rest/SearchResourceTest.java (hunk 5 above)
Comment (Runi-VN) [reply src/test/java/rest/SearchResourceTest.java]: I documented it on [line 150](https://github.com/MalteMagnussen/CA2/pull/53/files/d80fc10e8b9a9c20a4427408cf88dd59a6ea6aed#diff-a138706ed1fa2327b11739b17e647eb7R150) using the following stackoverflow link:
https://stackoverflow.com/a/49694288

Essentially we, with JUnit 5, can take use of `TestInfo`, an element that contains information about the test being run.

In our @BeforeEach (see above) I made some changes to take into account if the test is tagget with `skipBeforeEach` (line 420).

I essentially skip the @BeforeEach for the method `testGetAllZipCodes_Exception1()` because I need to test exception throwing for an empty database.

Comment (MalteMagnussen) [reply src/test/java/rest/SearchResourceTest.java]: Thanks, great explanation.

File: src/test/java/rest/SearchResourceTest.java (hunk 4 above)
Comment (MalteMagnussen) [reply src/test/java/rest/SearchResourceTest.java]: Thanks man. Nice. Where does that message go?

Comment (Runi-VN) [reply src/test/java/rest/SearchResourceTest.java]: It's primarily for logging I believe

Comment (MalteMagnussen) [reply src/test/java/rest/SearchResourceTest.java]: Ahh alright.

Comment (HrBjarup) [reply src/test/java/rest/SearchResourceTest.java]: Annoying that you had to do it this way but nice solution
